### ISSUE TITLE
Discussion: Restructuring the Integration Tests

### ISSUE THREAD

**Issue by oguzziya** [precice/precice#1021]
As our codebase and the application areas getting bigger, our integration tests are also getting bigger, which is really nice. However, currently, it is getting dirtier and dirtier both in terms of configuration files and the longer `SerialTests.cpp` and `ParallelTests.cpp` files. 

**Describe the solution you propose.**
As the first step, I suggest to move the integration tests into a separate folder, such as `src/precice/tests/config-files`. While doing so, we can fix a naming scheme for the configuration files as well. In addition, we can even remove duplicated configuration files and group them into seperate subfolders even.
 As the second step, which would be more involved, we can try to seperate the integration tests into seperate files under the same directory. It would be not as trivial as the unit tests due to the nature of the integration tests but it should be doable.

**Describe alternatives you've considered**
Do not touch at all, everyone already hates implementing tests.

**Additional context**
We can check other projects about how they are approaching this topic.

**Comment by uekerman** [comment]
Yes, I fully agree that we need to do sth here.

> As the first step, I suggest to move the integration tests into a separate folder, such as src/precice/tests/config-files

I guess you mean "integration test config files", right? If yes, then this is indeed a good step already.

> we can fix a naming scheme for the configuration files as well

yes :+1:

> and group them into seperate subfolders even.

Let's better go step by step.

> As the second step, which would be more involved, we can try to seperate the integration tests into seperate files under the same directory. It would be not as trivial as the unit tests due to the nature of the integration tests but it should be doable.

Some more structure would be good, yes. Big open question is how to structure them.

**Comment by fsimonis** [comment]
A fundamental question is: How do we group tests?

Currently we only group serial/parallel tests into two test suites and files. 
We use some tests suits beyond this, but this is rare.

Some ideas
```
# Group by feature, merge serial and parallel, tests go into tests.cpp.
# Prefix configs with the feature.

tests/
+ GlobalRBFPartitioning/
| + tests.cpp
| + GlobalRBFPartitioning.xml
+ QN/
| + tests.cpp
| + QN1.xml
| + QN2.xml
+ MeshRequirements/
  + tests.cpp
  + MeshRequirements-nn.xml
  + MeshRequirements-np.xml

# Group by serial/parallel, then feature. tests go into tests.cpp
# Configs have free names

tests/
+ Parallel/
| + GlobalRBFPartitioning/
| | + tests.cpp
| | + config.xml
| + QN/
|   + tests.cpp
|   + QN1.xml
|   + QN2.xml
+ Serial/
  + MeshRequirements/
  | + tests.cpp
  | + nn.xml
  | + np.xml 
  + Lifecycle/
    + tests.cpp
    + lifecycle.xml
```

Let's say the test suites should reflect the folder structure. In that case, you could run tests simply by looking at the path.

**Comment by BenjaminRodenberg** [comment]
In 8cfaef3 I started to put some of my new tests into a folder. This is currently a draft, but might be helpful as a basis for discussion, since it already goes a bit into the direction described by @fsimonis above. Feel free to review https://github.com/BenjaminRodenberg/precice/pull/1 and comment on the testing over there.

**Comment by fsimonis** [comment]
The logical grouping would be by "tested feature". or "scenario".
Tests testing multiple things should be split up into multiple tests.

We could also keep the old grouping in boost and group the tests in files by feature.
Then a file `QNTests.cpp` would include both `Serial` and `Parallel` test suites.

Refactoring our tests like this will also be a good exercise in understanding what we actually test.
@uekerman as you have written the majority of the integration tests, which approach looks best to you?

**Comment by BenjaminRodenberg** [comment]
Grouping by feature and also distributing the configuration files into folders sounds like a good idea to me. Following the structure from https://github.com/precice/precice/issues/1021#issuecomment-943297766, I have one more question, which is currently unanswered:

Do we want to allow deep hierarchies? If we group by feature, we might still get large `tests.cpp` files for each feature (see for example https://github.com/BenjaminRodenberg/precice/blob/ee666faa2aeae206583820afffe6c1aff29a6f2a/src/precice/tests/Time/TimeTests.cpp). I already made the experience that merging and reorganizing such large files is difficult.

We could even go one step further and use a folder for a suite and a single file for each individual test. This also forces us to properly factor out logic that covers multiple tests (fixtures, or [functions used by multiple tests](https://github.com/precice/precice/blob/b784b3849142693c315a16698328bfa975b348e5/src/precice/tests/SerialTests.cpp#L2329)). This might also help us to make fixtures and functions more visible and helps reusing them. Currently these things often get lost in the gigantic source files.

But we should also keep in mind that this will cause some synchronization overhead, because we have to keep folder/file names and names of suites/tests consistent.

**Comment by uekerman** [comment]
> Tests testing multiple things should be split up into multiple tests.

I am not to sure about this. In the end, we are talking about integration tests here. We want to test, for example, if the repartitioning still works correctly when we use a quasi-Newton method and an RBF mapping at the same time.

> We could also keep the old grouping in boost and group the tests in files by feature.

I think it would be easier to overview if boost structure and file structure are exactly consistent.

> We could even go one step further and use a folder for a suite and a single file for each individual test. 

This sounds over-engineered to me. Tests that belong together should in my opinion also be in the same file.

> This also forces us to properly factor out logic that covers multiple tests

Why? Wouldn't this make it more complicated?

> @uekerman as you have written the majority of the integration tests, which approach looks best to you?

Not sure if this is actually true :see_no_evil:

I guess the only reasonable way is to do some structure according to features or type of tests. If we end up with 5-7 categories instead of the two we have now (serial and parallel) we would already win quite some overview. Making the granularity too fine could hurt more than it helps. Even with 5-7 categories we will find tests that could fit in multiple categories. Then we simply have to pick one.

You wants to suggest a categorization? :grin:

**Comment by BenjaminRodenberg** [comment]
> > This also forces us to properly factor out logic that covers multiple tests
> 
> Why? Wouldn't this make it more complicated?

I'm mostly thinking about the following situation:

### Common logic

https://github.com/precice/precice/blob/d2d2e2b4d774ae75cafee4ab889efbc55124ef0e/src/precice/tests/SerialTests.cpp#L3745

### Test 1

https://github.com/precice/precice/blob/d2d2e2b4d774ae75cafee4ab889efbc55124ef0e/src/precice/tests/SerialTests.cpp#L3871-L3876

### Test 1

https://github.com/precice/precice/blob/d2d2e2b4d774ae75cafee4ab889efbc55124ef0e/src/precice/tests/SerialTests.cpp#L3878-L3883

### How could this look like?

If we apply the rule "every test has its own file" (`multiCoulplingFourSolvers1.cpp` and `multiCoulplingFourSolvers2.cpp` ), we have to move `multiCouplingFourSolvers(...)` into it's own file (`helpers.cpp` or similar) to be accessible for each of the two tests. This could result in the following folder structure:

```
serial/
└── multiCouplingFourSolvers
    ├── helpers.cpp
    ├── helpers.hpp
    ├── multi-coupling-four-solver-1.xml
    ├── multi-coupling-four-solver-2.xml
    ├── MultiCouplingFourSolvers1.cpp
    └── MultiCouplingFourSolvers2.cpp
```

We might get deep hierarchies and many files, but it also forces us to clean up includes, fixtures etc. I will try to push a draft PR implementing this example today. Currently I think the main problem is not that we have too many folders or files, but that we have files that are too long, contain a lot of duplicated or irrelevant code and the mapping of config files to tests is often unclear. Splitting the large files into small files helps us to clean this up for each individual test and we can also easily move around tests without getting merge conflicts or hard-to-read diffs.

**Comment by BenjaminRodenberg** [comment]
Please take a look at c0acf71 for an example implementation.

**Comment by BenjaminRodenberg** [comment]
I'm trying to summarize the options and opinions mentioned in https://github.com/precice/precice/issues/1021#issuecomment-974861728 here to get a better overview.

1. A `BOOST_AUTO_TEST_CASE` testing multiple things should be split up into multiple `BOOST_AUTO_TEST_CASE`s (not necessarily equal to files!):
    1. Yes: @fsimonis  
    2. No: @uekerman, @BenjaminRodenberg
2. When to use a folder below `src/precice/tests/`:
    1. Folder structure equals structure of boost test suites (#1140 at 457abf4) @BenjaminRodenberg 
    2. Let's say the test suites should reflect the folder structure. In that case, you could run tests simply by looking at the path. @uekerman @fsimonis 
3. When to use an individual file:
    1. I think it would be easier to overview if boost structure and file structure are exactly consistent. @uekerman 
    2. We could also keep the old grouping in boost and group the tests in files by feature. @fsimonis 
    3. Tests that belong together should in my opinion also be in the same file. @uekerman 
    4. Every test has its own file (#1140 at 457abf4): @BenjaminRodenberg 
4. Where to put additional logic:
    1. Always into a `helpers.cpp` file, never into the test file (#1140 at 457abf4): @BenjaminRodenberg 
    2. Always into the file of the test using it: 

If I distribute votes incorrectly or forget something, feel free to edit. Sometimes I was also unsure about the actual opinion. So please just add your name, if you think an option sounds good to you. Please don't edit the existing options, if something is missing, you can append it. We can just create a new list, when this one becomes outdated.

**Comment by BenjaminRodenberg** [comment]
Some thoughts on https://github.com/precice/precice/issues/1021#issuecomment-975394469:

* On 2:
     * I think we are agreeing on the following point: Every already existing suite should have a folder. 
     * Open question: When should we create new suites? Should we go deeper than `src/precice/tests/serial` and `src/precice/tests/parallel`?
     * It might be a good idea to implement the folder hierarchy following the already existing suites, if everybody agrees here. This is a step into the right direction.
* I am actually not sure where to draw the line between 3.i., 3.ii, 3.iii

**Comment by fsimonis** [comment]
@BenjaminRodenberg and I had a productive in-person discussion regarding this matter.

We came to the following conclusion
* Move integration tests out of `src/precice/` and put them into `tests/`.  
* Group by feature, nesting of suites allowed
* Suite nesting = path structure
* 1 test per cpp file
* test name = test file name = test config name
* put helpers into file `helpers.[hc]pp`, located as close the test file as possible.

## Example
```
tests/Serial/MultiCouplingFourSolvers/MultiCouplingFourSolvers1.cpp
tests/Serial/MultiCouplingFourSolvers/MultiCouplingFourSolvers1.xml
```

# Bonus:
This scheme allows to automatically generate:
* The directory structure
* The test implementation and configuration 

Example:
```
$ python3 create_test.py -s /Serial/MultiCouplingFourSolvers -t MultiCouplingFourSolvers1
mkdir -p tests/Serial/MultiCouplingFourSolvers
generate tests/Serial/MultiCouplingFourSolvers/MultiCouplingFourSolvers1.cpp
generate tests/Serial/MultiCouplingFourSolvers/MultiCouplingFourSolvers1.xml
```

How to migrate to the new scheme:

First step:
1. Move everything to `tests/`
2. Group tests and configs into `tests/Serial/` and `tests/Parallel/`
3. Move includes into helper files
4. Split all tests into separate `cpp` files
5. Fix updateSourceFiles tool
6. Fix fixtures
6. Get it all running

Second step:
1. regroup tests by features

**Comment by BenjaminRodenberg** [comment]
I tried to implement the proposed steps from https://github.com/precice/precice/issues/1021#issuecomment-976698077 for a single test case in #1140. There are some side effects of moving the integration tests from `precice/src/precice/tests` to `precice/tests`. I was at least not aware of these effects from the beginning:

* Do have to update `CMakeLists.txt`?
* We have to update `tools/formatting/format-all`.
* We have to update `tools/building/updateSourceFiles.py`
* More similar scripts?
* Not sure whether `precice/src/tests.cmake` is still the ideal place, now that we have files in `precice/tests`?

@fsimonis can you take a look at #1140 and fix and/or comment? I am not really very experienced with `cmake` and all these scripts and would need some input. These hurdles might also be an argument to keep the tests where they are and remove the point *Move everything to `tests/`* from above.

**Comment by fsimonis** [comment]
Side note: It is possible to get the names of all parent suites of a test.
Meaning that we can generate the full path of the test location at runtime.

This allows us to generate the complete configuration file path and name.
We could even check if the config exists.
```cpp
// file: ./tests/Serial/QN/Subcycling/Example.cpp

BOOST_AUTO_TEST_SUITE(Serial)
BOOST_AUTO_TEST_SUITE(QN)
BOOST_AUTO_TEST_SUITE(Subcycling)

BOOST_TEST(Example) {
PRECICE_TEST("A"_on(1_rank), "B"_on(2_ranks));
  SolverInterface interface(context.rank, context.size, context.config());
  // looks for ./tests/Serial/QN/Subcycling/Example.xml
 
  // ...
}

// ...
```

**Comment by uekerman** [comment]
I agree with the refactoring suggestions. I think it's going in the right direction. Only leaving some remarks here.

* Drawback from moving the integration tests to `/tests`: users and future developers might think that these are all our tests, overlooking the unit tests. Not a reasons against the moving, but we have to be aware of this and document carefully.
* I like the auto-generation idea of tests. `create_test.py` could even be a part of the future controller. There, we generate a preCICE configuration from a coupling topology description. Generating the `MultiCouplingFourSolvers1.cpp` on top, we could then already fill the `if (context.isNamed("SolverOne"))` etc. But that's future work.
* "regroup tests by features" will still be the complicated and important step. Please don't get blinded by all the other refactoring work :grin:

**Comment by BenjaminRodenberg** [comment]
>     * I like the auto-generation idea of tests. `create_test.py` could even be a part of the future controller. There, we generate a preCICE configuration from a coupling topology description. Generating the `MultiCouplingFourSolvers1.cpp` on top, we could then already fill the `if (context.isNamed("SolverOne"))` etc. But that's future work.

Then I will try to provide a first version for the autogeneration tool in #1140.

**Comment by uekerman** [comment]
> Then I will try to provide a first version for the autogeneration tool in #1140.

Generating `precice-config.xml` is way too complicated :see_no_evil: for a quick hit. Let's look into this after the complete refactoring.

**Comment by fsimonis** [comment]
I noticed that we currently use `PreciceTests` as a common test suite for integration tests.

Could we please rename this to something less offensive to our project?
I propose the simply `Integration` suite.

**Comment by uekerman** [comment]
Sure, `IntegrationTests` or just `Integration`?

**Comment by fsimonis** [comment]
```
$ mpirun -n 4 ./testprecice --run_test=IntegrationTests/QN/Subcycling
$ mpirun -n 4 ./testprecice --run_test=Integration/QN/Subcycling
```
@uekerman Maybe the shorter version?

**Comment by uekerman** [comment]
"Integration" alone could mean anything. Will there always be context to implicitly get the "tests"?
If yes, then shorter is better.

**Comment by BenjaminRodenberg** [comment]
Closing this issue. I just stumbled over it (#1561). I think the basic goal of this issue was solved a long while ago via #1205

**Comment by uekerman** [comment]
I disagree :grin: 
We did not yet do the actual restructuring, no?

**Comment by BenjaminRodenberg** [comment]
> I disagree grin We did not yet do the actual restructuring, no?

Then the title of #1205 is at least very very misleading :see_no_evil: 

@uekerman what kind of restructuring are we missing? I guess you mean the difficult part with finding the right names and folders for the tests and not "just" rearranging things :smirk:

**Comment by uekerman** [comment]
> I guess you mean the difficult part with finding the right names and folders for the tests and not "just" rearranging things

Yes. So far, we developed the tooling and some conventions, but we have not yet sorted out the tests.
